File: domains/integration-test/helpers/noto_helper.go
/*
 * Copyright © 2024 Kaleido, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package helpers

import (
	"context"
	_ "embed"
	"testing"

	"github.com/LFDT-Paladin/paladin/domains/noto/pkg/types"
	"github.com/LFDT-Paladin/paladin/sdk/go/pkg/pldtypes"
	"github.com/LFDT-Paladin/paladin/sdk/go/pkg/rpcclient"
	"github.com/LFDT-Paladin/paladin/sdk/go/pkg/solutils"
	"github.com/hyperledger/firefly-signer/pkg/abi"
	"github.com/stretchr/testify/require"
)

//go:embed abis/NotoFactory.json
var NotoFactoryJSON []byte

//go:embed abis/Noto.json
var NotoJSON []byte

//go:embed abis/Noto_V0.json
var NotoV0JSON []byte

//go:embed abis/ERC1967Proxy.json
var ERC1967ProxyJSON []byte

//go:embed abis/INoto.json
var NotoInterfaceJSON []byte

//go:embed abis/INoto_V0.json
var NotoV0InterfaceJSON []byte

var NotoFactoryABI = solutils.MustParseBuildABI(NotoFactoryJSON)

type NotoHelper struct {
	t       *testing.T
	rpc     rpcclient.Client
	Address *pldtypes.EthAddress
	ABI     abi.ABI
}

func DeployNoto(ctx context.Context, t *testing.T, rpc rpcclient.Client, domainName, notary string, hooks *pldtypes.EthAddress) *NotoHelper {
	return DeployNotoImplementation(ctx, t, rpc, domainName, "", notary, hooks)
}

func DeployNotoImplementation(ctx context.Context, t *testing.T, rpc rpcclient.Client, domainName, implementation, notary string, hooks *pldtypes.EthAddress) *NotoHelper {
	notaryMode := types.NotaryModeBasic
	if hooks != nil {
		notaryMode = types.NotaryModeHooks
	}

	var addr pldtypes.EthAddress
	constructorParams := &types.ConstructorParams{
		Notary:     notary + "@node1",
		NotaryMode: notaryMode,
		Options: types.NotoOptions{
			Hooks: &types.NotoHooksOptions{
				PublicAddress:     hooks,
				DevUsePublicHooks: true,
			},
		},
	}
	if implementation != "" {
		constructorParams.Implementation = implementation
		constructorParams.Name = implementation // This shouldn't be necessary but it's a workaround for a bug in the factory
	}
	rpcerr := rpc.CallRPC(ctx, &addr, "testbed_deploy", domainName, "notary", constructorParams)
	if rpcerr != nil {
		require.NoError(t, rpcerr)
	}
	return &NotoHelper{
		t:       t,
		rpc:     rpc,
		Address: &addr,
		ABI:     solutils.MustLoadBuild(NotoInterfaceJSON).ABI,
	}
}

func (n *NotoHelper) Mint(ctx context.Context, to string, amount int64) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["mint"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, &types.MintParams{
		To:     to,
		Amount: pldtypes.Int64ToInt256(amount),
	}))
}

func (n *NotoHelper) Transfer(ctx context.Context, to string, amount int64) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["transfer"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, &types.TransferParams{
		To:     to,
		Amount: pldtypes.Int64ToInt256(amount),
	}))
}

func (n *NotoHelper) TransferFrom(ctx context.Context, from, to string, amount int64) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["transferFrom"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, &types.TransferFromParams{
		From:   from,
		To:     to,
		Amount: pldtypes.Int64ToInt256(amount),
	}))
}

func (n *NotoHelper) Burn(ctx context.Context, amount int64) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["burn"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, &types.BurnParams{
		Amount: pldtypes.Int64ToInt256(amount),
	}))
}

func (n *NotoHelper) BurnFrom(ctx context.Context, from string, amount int64) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["burnFrom"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, &types.BurnFromParams{
		From:   from,
		Amount: pldtypes.Int64ToInt256(amount),
	}))
}

func (n *NotoHelper) ApproveTransfer(ctx context.Context, params *types.ApproveParams) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["approveTransfer"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, params))
}

func (n *NotoHelper) Lock(ctx context.Context, params *types.LockParams) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["lock"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, params))
}

func (n *NotoHelper) Unlock(ctx context.Context, params *types.UnlockParams) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["unlock"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, params))
}

func (n *NotoHelper) PrepareUnlock(ctx context.Context, params *types.PrepareUnlockParams) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["prepareUnlock"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, params))
}

func (n *NotoHelper) DelegateLock(ctx context.Context, params *types.DelegateLockParams) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["delegateLock"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, params))
}

func (n *NotoHelper) BalanceOf(ctx context.Context, params *types.BalanceOfParam) *DomainTransactionHelper {
	fn := types.NotoABI.Functions()["balanceOf"]
	return NewDomainTransactionHelper(ctx, n.t, n.rpc, n.Address, fn, toJSON(n.t, params))
}

File: domains/pente/src/test/java/io/kaleido/paladin/pente/domain/helpers/NotoHelper.java
/*
 * Copyright © 2024 Kaleido, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
 * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations under the License.
 *
 * SPDX-License-Identifier: Apache-2.0
 */

package io.kaleido.paladin.pente.domain.helpers;

import com.fasterxml.jackson.databind.JsonNode;
import io.kaleido.paladin.testbed.Testbed;

import com.fasterxml.jackson.annotation.JsonIgnoreProperties;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.databind.ObjectMapper;
import io.kaleido.paladin.pente.domain.PenteConfiguration.GroupTupleJSON;
import io.kaleido.paladin.toolkit.*;

import java.io.IOException;
import java.util.ArrayDeque;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class NotoHelper {
    final String domainName;
    final Testbed testbed;
    final JsonABI abi;
    final String address;

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record ConstructorParams(
            @JsonProperty
            String notary,
            @JsonProperty
            String notaryMode,
            @JsonProperty
            OptionsParams options
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record OptionsParams(
            @JsonProperty
            HookParams hooks
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record HookParams(
            @JsonProperty
            String publicAddress,
            @JsonProperty
            JsonHex.Address privateAddress,
            @JsonProperty
            GroupTupleJSON privateGroup
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record NotoCoin(
            @JsonProperty
            NotoCoinData data
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record NotoCoinData(
            @JsonProperty
            String owner,
            @JsonProperty
            String amount
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record ReceiptLockInfo(
            @JsonProperty
            JsonHex.Bytes32 lockId,
            @JsonProperty
            JsonHex.Address delegate,
            @JsonProperty
            JsonNode unlockParams,
            @JsonProperty
            JsonHex.Bytes unlockCall
    ) {
    }

    @JsonIgnoreProperties(ignoreUnknown = true)
    public record NotoDomainReceipt(
            @JsonProperty
            ReceiptLockInfo lockInfo
    ) {
    }

    public static NotoHelper deploy(String domainName, String from, Testbed testbed, ConstructorParams params) throws IOException {
        String address = testbed.getRpcClient().request("testbed_deploy", domainName, from, params);
        JsonABI abi = JsonABI.fromJSONResourceEntry(
                NotoHelper.class.getClassLoader(),
                "contracts/domains/interfaces/INotoPrivate.sol/INotoPrivate.json",
                "abi"
        );
        return new NotoHelper(domainName, testbed, abi, address);
    }

    private NotoHelper(String domainName, Testbed testbed, JsonABI abi, String address) {
        this.domainName = domainName;
        this.testbed = testbed;
        this.abi = abi;
        this.address = address;
    }

    public String address() {
        return address;
    }

    public List<NotoCoin> queryStates(JsonHex.Bytes32 schemaID, JsonQuery.Query query) throws IOException {
        List<HashMap<String, Object>> states = testbed.getRpcClient().request("pstate_queryContractStates",
                domainName, address, schemaID, query, "available");
        var mapper = new ObjectMapper();
        return states.stream().map(state -> mapper.convertValue(state, NotoCoin.class)).toList();
    }

    public void mint(String sender, String to, int amount) throws IOException {
        testbed.getRpcClient().request("testbed_invoke", new Testbed.TransactionInput(
                "private",
                "",
                sender,
                JsonHex.addressFrom(address),
                new HashMap<>() {{
                    put("to", to);
                    put("amount", amount);
                    put("data", "0x");
                }},
                abi,
                "mint"
        ), true);
    }

    public void transfer(String sender, String to, int amount) throws IOException {
        testbed.getRpcClient().request("testbed_invoke", new Testbed.TransactionInput(
                "private",
                "",
                sender,
                JsonHex.addressFrom(address),
                new HashMap<>() {{
                    put("to", to);
                    put("amount", amount);
                    put("data", "0x");
                }},
                abi,
                "transfer"
        ), true);
    }

    public Testbed.TransactionResult lock(String sender, int amount) throws IOException {
        return TestbedHelper.getTransactionResult(
                testbed.getRpcClient().request("testbed_invoke", new Testbed.TransactionInput(
                        "private",
                        "",
                        sender,
                        JsonHex.addressFrom(address),
                        new HashMap<>() {{
                            put("amount", amount);
                            put("data", "0x");
                        }},
                        abi,
                        "lock"
                ), true));
    }

    public Testbed.TransactionResult prepareUnlock(String sender, JsonHex.Bytes32 lockId, String from, String to, int amount) throws IOException {
        return TestbedHelper.getTransactionResult(
                testbed.getRpcClient().request("testbed_invoke", new Testbed.TransactionInput(
                        "private",
                        "",
                        sender,
                        JsonHex.addressFrom(address),
                        new HashMap<>() {{
                            put("lockId", lockId);
                            put("from", from);
                            put("recipients", new ArrayDeque<Map<String, Object>>() {{
                                add(new HashMap<>() {{
                                    put("to", to);
                                    put("amount", amount);
                                }});
                            }});
                            put("data", "0x");
                        }},
                        abi,
                        "prepareUnlock"
                ), true));
    }

    public void delegateLock(String sender, JsonHex.Bytes32 lockId, JsonHex.Address delegate, JsonNode unlockParams) throws IOException {
        testbed.getRpcClient().request("testbed_invoke", new Testbed.TransactionInput(
                "private",
                "",
                sender,
                JsonHex.addressFrom(address),
                new HashMap<>() {{
                    put("lockId", lockId);
                    put("unlock", unlockParams);
                    put("delegate", delegate);
                    put("data", "0x");
                }},
                abi,
                "delegateLock"
        ), true);
    }
}
